### ISSUE TITLE
OBB (Oriented Binding Box) coordinate format conversion method

### ISSUE THREAD

**Issue by haaaaak** [ultralytics/ultralytics#7103]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello! I am working on OBB code. Among them, I am asking you a question because I wonder how the data of the OBB model is processed.
The structure of the OBB data was found to have the same format as segmentation task, such as x1, y1, x2, y2, ... and .
I checked this information and conducted OBB learning with the coordinate data of the segmentation format, however, in the training process and detection process, it was confirmed that the coordinates were treated as bounding boxes.
and so, I thought that the transformation was made in the xyxyxy2xywhr function in ultralytics/utils/ops.py

```
def xyxyxyxy2xywhr(corners):
    is_torch = isinstance(corners, torch.Tensor)
    points = corners.cpu().numpy() if is_torch else corners
    points = points.reshape(len(corners), -1, 2)
    rboxes = []
    for pts in points:
        # NOTE: Use cv2.minAreaRect to get accurate xywhr,
        # especially some objects are cut off by augmentations in dataloader.
        (x, y), (w, h), angle = cv2.minAreaRect(pts)
        rboxes.append([x, y, w, h, angle / 180 * np.pi])
    rboxes = torch.tensor(rboxes, device=corners.device, dtype=corners.dtype) if is_torch else np.asarray(
        rboxes, dtype=points.dtype)
    return rboxes
```

However, when I backtracked the OBB train code, there was no part using xyxyxyxy2xywhr.
In theory, the angle of rotation cannot be obtained from coordinate data alone.
 How can the coordinates and angle of rotation of bounding boxes be obtained from learning or detect?

In addition, I wonder if it is not a problem to use a data format with more coordinates than x1, y1, ..., x4, y4 as Training data in OBB.

I look forward to hearing from you, and thank you!

**Comment by glenn-jocher** [comment]
@haaaaak hello! Thanks for reaching out with your inquiry. 🌟

The OBB (Oriented Bounding Box) training and detection processes in YOLOv8 indeed handle data differently compared to the segmentation tasks. OBBs include rotation information, which is not captured by using standard axis-aligned bounding boxes or segmentation polygons.

For training or detecting with OBBs, the model requires an additional rotation parameter, beyond the simple corner coordinates (x1, y1, ..., x4, y4). The `xyxyxyxy2xywhr` function you mentioned converts the corner points to a central point (x, y), width (w), height (h), and rotation (r), which is crucial for proper OBB handling.

However, if you did not find the use of `xyxyxyxy2xywhr` in the training code, it suggests that rotation parameters may need to be supplied differently or are managed internally in a way that is not immediately evident from the code you've inspected.

As for training with more coordinates, YOLOv8 expects a specific format for OBB inputs. If you input data with more than the expected number of coordinates (x1, y1, ..., x4, y4), the extra points may be ignored or can cause errors during training and inference as they don't conform to the expected OBB format.

For the specifics on incorporating the rotation angles and managing the coordinate conversion effectively for training and detection, I would suggest checking out the comprehensive guidelines in our documentation under the `Train`, `Val`, and `Predict` modes. Insight on how YOLOv8 handles such cases will be crucial to thoroughly understanding and implementing the OBB model you're working on.

If you have further questions or if there’s anything else my team and I can do to assist you, please feel free to ask. We appreciate your interest in our work and are here to help! Happy coding! 🚀

**Comment by haaaaak** [comment]
First of all, thank you for your answer! 😁

What I'm curious about is that the format of the data label to utilize the OBB model has this structure, as far as I know, class_id x1 y1 x2 x3 x4 y4
And, no bounding box angle information exists in this information.
I would appreciate it if you could tell me where the angle of the bounding box is obtained in the training process and the detection process.

In addition, there is a problem that the txt label is not saved even if save_txt is specified as True.

`
result = model('datasets/seg_test/images/test', imgsz=1024, conf=0.25, save=True, show_labels=False, save_txt=True)
`

```
...
Spped: 3.7ms preprocess, 3.4ms inference, 1.2ms postprocess per image at shape(1, 3, 1024, 1024)
Results saved to runs/obb/predict4
0 label saved to runs/obb/predict4/labels
```

I actually checked the folder, but the label did not exist.

Please answer where the angle information is obtained in the OBB format, and the problem that the label information is not saved even if save_txt=True

**Comment by glenn-jocher** [comment]
@haaaaak, glad to help! 😊

OBB model training and detection inherently require the angle information for processing. However, YOLOv8's current public repository does not natively support OBB formats with rotation angles directly in the label files. The angle for an oriented bounding box typically requires explicit annotation, which may not be present in the standard labeling formats.

For the OBB angle, a custom function or preprocessing step would be necessary to calculate and include this angle during training and detection. This is not a built-in feature, and implementing OBB handling would typically require modifying the YOLOv8 codebase or preprocessing your data accordingly.

Regarding your issue with `save_txt=True` not saving the label files, there could be several reasons for this. Please ensure that:
- The images have detectable objects with confidences above the specified threshold.
- The result directory (`runs/obb/predict4`) has the correct write permissions.
- The dataset is correctly labeled and formatted for detection.

If you've verified the above points and are still facing issues, please check the documentation on `Predict` mode for further guidance, or provide specific details about the error, so we can assist you more effectively.

Continued success with your OBB project!

**Comment by haaaaak** [comment]
@glenn-jocher So the YOLOv8 OBB model doesn't use any angle information during training or prediction?
This image is part of DOTAv2's data set, and as you can see, there is no labeling data that exists at right angles in the image.
I'm curious about how coordinates are processed in learning or predicting from these data.

![capture](https://github.com/ultralytics/ultralytics/assets/97625003/2a465a95-4151-4952-b6f0-998d1ba0a221)

What I want to ask is exactly what format these coordinates enter when learning. Do it not use any information about angles?

**Comment by glenn-jocher** [comment]
@haaaaak, in YOLOv8, we currently do not handle OBBs with explicit rotation angles for training or prediction in the standard workflow. If your dataset, such as DOTAv2, contains annotations for rotated objects without explicit angle values, the model will not interpret these as OBBs, but rather as axis-aligned bounding boxes or polygons.

For OBB handling with angles, a customized approach is necessary, which is not part of the out-of-the-box functionality we offer. Training and predicting with OBBs and angle data would require bespoke pre-processing and network adjustments to interpret and learn from the rotated bounding box format. 🔄

The coordinates from the image dataset are used to generate either polygons or axis-aligned bounding boxes during the learning process. Angle information is not directly utilized unless custom modifications to the codebase are performed.

**Comment by kaunghtetsan275** [comment]
Hello. I have similar question. How does YOLO-OBB internally convert the annotation format? Because when I inspect the code, the function `xyxyxyxy2xywhr` is called from the `__call__` function of Format class, but I think `__call__` function isn't used or I'm not used when it is used. 
![Screenshot 2024-02-06 221907](https://github.com/ultralytics/ultralytics/assets/21242101/0ece507a-aeca-4bfc-b959-cad91fe03850)
But, the obb training works somehow. When I tried to predict using yolo-obb model, it can predict successfully and I have obb (`xyxyxyxy`) format in Result object. So, I believe the model uses `xywhr` instead of `xyxyxyxy` during training. My question might be similar to @haaaaak but I really want to know how the training handle and process dataset annotation. Could you please enlighten me on this issue? Thank you.

**Comment by glenn-jocher** [comment]
@kaunghtetsan275 hello! YOLOv8 does not natively support OBB annotations with rotation angles in the standard training or prediction processes. The `xyxyxyxy2xywhr` function you've found is not part of the typical YOLOv8 workflow and might be a remnant from a custom implementation or an experimental feature not officially supported.

During training, YOLOv8 uses axis-aligned bounding boxes without rotation angles. If you're seeing OBB predictions (`xyxyxyxy` format), it suggests there might be custom code handling this within your environment. The model itself is designed to output standard bounding boxes, and any conversion to OBB would require additional steps outside the standard YOLOv8 pipeline.

For a deeper understanding of how annotations are handled, you would need to review the custom code or modifications made to the YOLOv8 repository that you're working with. The official YOLOv8 documentation and codebase do not include OBB handling with rotation angles.

**Comment by kaunghtetsan275** [comment]
Thanks. Let me clarify my question again. I found YOLOv8 has a specific task called "obb" (#4499, [obb docs](https://github.com/ultralytics/ultralytics/blob/main/docs/en/tasks/obb.md))  just like "pose" and "segmentation" but for oriented bounding box prediction. I tried it out and found that issue I've previously mentioned. After checking the code again, `xyxyxyxy2xywhr` method was indeed called from data augmentation during obb training, and YOLOv8's OBB head predicts the angle `r` in addition to the box `xywh`.
![Screenshot 2024-02-08 103952](https://github.com/ultralytics/ultralytics/assets/21242101/6f75718f-b62d-46ea-aea9-468169c7807d)
I was trying to modify the annotation conversion as part of my master thesis. I'm now trying to understand more about how the head and loss function works. It would be really helpful if there is a documentation related to how they work. I found [this doc](https://github.com/ultralytics/ultralytics/blob/main/docs/en/reference/nn/modules/head.md) regarding head but it is still incomplete. Anyhow, I have learned a lot from well-written code and documentation. Please keep up the good work!

**Comment by glenn-jocher** [comment]
@kaunghtetsan275 hello! Thanks for reaching out with your question about Oriented Bounding Box (OBB) coordinate format conversion in YOLOv8.

The `xyxyxyxy2xywhr` function you've found in `ultralytics/utils/ops.py` is indeed used to convert 8-point polygon coordinates (x1, y1, x2, y2, ..., x4, y4) into a rotated bounding box format (x_center, y_center, width, height, rotation_angle). This conversion is necessary because YOLOv8's OBB implementation requires the angle of rotation to predict oriented bounding boxes.

The conversion function uses OpenCV's `cv2.minAreaRect`, which calculates the minimum-area bounding rectangle for a set of points, allowing us to obtain the rotation angle. This is used during the training and detection processes to handle OBBs.

Regarding your question about using more coordinates than the standard 8 for training data in OBB, it's important to adhere to the expected format. YOLOv8's OBB expects exactly 8 coordinates representing the four corners of the bounding box. Using more coordinates could lead to unexpected behavior or errors, as the model is designed to work with quadrilateral OBBs.

If you have additional coordinates or more complex shapes, you might need to preprocess your data to fit the OBB format or consider using segmentation if the shapes cannot be accurately represented by OBBs.

I hope this clarifies your concerns! If you have further questions or need more assistance, feel free to ask. 😊👍
